### PR TITLE
feat(model): add mnist fsdp train script

### PR DESCRIPTION
**Pull request by Major-333**
The dlrover system will support elastic training under the FSDP strategy in the future. This PR will provide a beginner-level example of FSDP training, which uses CNN to train a recognition task on the MNIST dataset. The sample code is located in: `model_zoo/pytorch/fsdp_examples/mnist.py`. This code is based on `model_zoo/pytorch/mnist_cnn.py`, but with some modifications to support the FSDP training strategy.

I conducted experiments on a k8s cluster made up of two machines, which are respectively equipped with 4 * 2080Ti and 8 * 3090 graphics cards. I used replica: 12 for the experiment, and the training process is as shown in the following figure:

In addition, the logs for worker0 and worker1 are as follows:

- worker0: 
```
[2023-07-25 13:14:35,770] [INFO] [master_client.py:612:build_master_client] Build master client to connect with the DLRover master.
[2023-07-25 13:14:35,773] [INFO] [master_client.py:76:__init__] dlrover master addr is elasticjob-torch-mnist-gpu-28-dlrover-master:50001
[2023-07-25 13:14:35,775] [INFO] [training.py:433:launch_agent] Starting elastic_operator with launch configs:
  entrypoint       : model_zoo/pytorch/fsdp_examples/mnist.py
  min_nodes        : 12
  max_nodes        : 12
  nproc_per_node   : 1
  run_id           : none
  rdzv_backend     : static
  rdzv_endpoint    : 127.0.0.1:29500
  rdzv_configs     : {'rank': 0, 'timeout': 900, 'node_unit': 1}
  max_restarts     : 1
  monitor_interval : 5
  log_dir          : None
  metrics_cfg      : {}

[2023-07-25 13:14:35,780] [INFO] [training.py:333:_invoke_run] [default] starting workers for entrypoint: python
[2023-07-25 13:14:35,781] [INFO] [training.py:123:next_rendezvous] The node node_name attempts to join the next round of the rendezvous 'elastic-training' with timeout 600.
[2023-07-25 13:14:59,822] [INFO] [training.py:138:next_rendezvous] The nodetorch-mnist-gpu-28-edljob-worker-0 has joined round 0 of the elastic-training rendezvous as rank 0 in a world of size 12.
[2023-07-25 13:14:59,827] [INFO] [training.py:242:_rendezvous] [default] Rendezvous complete for workers. Result:
  restart_count=0
  master_addr=10.244.0.64
  master_port=54423
  group_rank=0
  group_world_size=12
  local_ranks=[0]
  role_ranks=[0]
  global_ranks=[0]
  role_world_sizes=[12]
  global_world_sizes=[12]

rank 0 is initialized local_rank = 0
Running basic FSDP example on local rank 0.
loss = 2.3013620376586914, step = 0
loss = 0.26035013794898987, step = 20
loss = 0.2508194148540497, step = 40
loss = 0.2582657039165497, step = 60
loss = 0.157688170671463, step = 80
loss = 0.3064134418964386, step = 100
loss = 0.20976434648036957, step = 120
loss = 0.14584095776081085, step = 140
Test model after epoch 0
Test the model ...

Test set: Average loss: 0.0774, Accuracy: 9759/10000 (98%)

loss = 0.1670873612165451, step = 0
loss = 0.23492978513240814, step = 20
loss = 0.3997991383075714, step = 40
loss = 0.13198073208332062, step = 60
loss = 0.23160065710544586, step = 80
loss = 0.08175430446863174, step = 100
loss = 0.07629987597465515, step = 120
loss = 0.12136860191822052, step = 140
Test model after epoch 1
Test the model ...
```
- worker1: 

```
[2023-07-25 13:14:37,029] [INFO] [master_client.py:612:build_master_client] Build master client to connect with the DLRover master.
[2023-07-25 13:14:37,032] [INFO] [master_client.py:76:__init__] dlrover master addr is elasticjob-torch-mnist-gpu-28-dlrover-master:50001
[2023-07-25 13:14:37,036] [INFO] [training.py:433:launch_agent] Starting elastic_operator with launch configs:
  entrypoint       : model_zoo/pytorch/fsdp_examples/mnist.py
  min_nodes        : 12
  max_nodes        : 12
  nproc_per_node   : 1
  run_id           : none
  rdzv_backend     : static
  rdzv_endpoint    : 127.0.0.1:29500
  rdzv_configs     : {'rank': 0, 'timeout': 900, 'node_unit': 1}
  max_restarts     : 1
  monitor_interval : 5
  log_dir          : None
  metrics_cfg      : {}

[2023-07-25 13:14:37,039] [INFO] [training.py:333:_invoke_run] [default] starting workers for entrypoint: python
[2023-07-25 13:14:37,042] [INFO] [training.py:123:next_rendezvous] The node node_name attempts to join the next round of the rendezvous 'elastic-training' with timeout 600.
[2023-07-25 13:15:01,081] [INFO] [training.py:138:next_rendezvous] The nodetorch-mnist-gpu-28-edljob-worker-1 has joined round 0 of the elastic-training rendezvous as rank 1 in a world of size 12.
[2023-07-25 13:15:01,084] [INFO] [training.py:242:_rendezvous] [default] Rendezvous complete for workers. Result:
  restart_count=0
  master_addr=10.244.0.64
  master_port=54423
  group_rank=1
  group_world_size=12
Running basic FSDP example on local rank 0.
loss = 2.3195579051971436, step = 0
loss = 0.920741081237793, step = 20
loss = 0.3812876045703888, step = 40
loss = 0.6321004629135132, step = 60
loss = 0.10630982369184494, step = 80
loss = 0.24280764162540436, step = 100
loss = 0.09044546633958817, step = 120
loss = 0.13473713397979736, step = 140
Test model after epoch 0
Test the model ...

Test set: Average loss: 0.0774, Accuracy: 9759/10000 (98%)

loss = 0.2743786573410034, step = 0
loss = 0.0632578507065773, step = 20
loss = 0.1030924990773201, step = 40
loss = 0.16128504276275635, step = 60
loss = 0.1300361305475235, step = 80
loss = 0.02153501659631729, step = 100
loss = 0.12761962413787842, step = 120
loss = 0.15265142917633057, step = 140
Test model after epoch 1
Test the model ...
```

In the future, we plan to introduce another example, utilizing the Fastchat repository, to demonstrate full-parameter fine-tuning of the llama model using FSDP.

### CODE REVIEW

File: model_zoo/pytorch/fsdp_examples/mnist.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
+# Copyright 2023 The DLRover Authors. All rights reserved.
```
Comment (workingloong) [inline model_zoo/pytorch/fsdp_examples/mnist.py:1]: model_zoo/pytorch/fsdp_examples/mnist.py -> model_zoo/pytorch/fsdp/mnist.py

File: model_zoo/pytorch/fsdp_examples/mnist.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    test_loader = DataLoader(dataset=test_dataset, batch_size=args.batch_size)
+
+    model = Net()
+
+    if torch.cuda.is_available():
+        local_rank = int(os.environ["LOCAL_RANK"])
+        print(f"Running basic FSDP example on local rank {local_rank}.")
+        # shared model
+        model = model.to(local_rank)
+        model = FSDP(model)
```
Comment (workingloong) [inline model_zoo/pytorch/fsdp_examples/mnist.py:123]: We don't need to create a new Python file. We can use an argument to use FSDP. For example:
```
if args.use_fdsp:
    model = FSDP(model)
```

Comment (Major-333) [reply model_zoo/pytorch/fsdp_examples/mnist.py:123]: 好的

File: dlrover/examples/torch_mnist_fsdp_job.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  replicaSpecs:
+    worker:
+      replicas: 4
+      template:
+        spec:
+          restartPolicy: Never
+          containers:
+            - name: main
+              # yamllint disable-line rule:line-length
+              image: registry.cn-hangzhou.aliyuncs.com/major333/dlrover:fsdp-test
```
Comment (merlintang) [inline dlrover/examples/torch_mnist_fsdp_job.yaml:18]: major333 is personal address?

**Review by merlintang — COMMENTED** [review]
does this related to all educe model training? if it is, can we add the all reduce model in the pr title?

File: model_zoo/pytorch/fsdp_examples/mnist.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
+# Copyright 2023 The DLRover Authors. All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
```
Comment (workingloong) [inline model_zoo/pytorch/fsdp_examples/mnist.py:6]: We'd better to remove the file.

File: dlrover/examples/torch_mnist_fsdp_job.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+              imagePullPolicy: IfNotPresent
+              command:
+                - /bin/bash
+                - -c
+                - >-
+                  dlrover-run --nnodes=$WORKER_NUM --nproc_per_node=1 \
+                  --max_restarts=1 model_zoo/pytorch/fsdp_examples/mnist.py \
```
Comment (workingloong) [inline dlrover/examples/torch_mnist_fsdp_job.yaml:25]: model_zoo/pytorch/fsdp_examples/mnist_cnn.py --use_fsdp

**Pull request merged**